app/services/db/proofing_component/add.rb

```diff
@@ -1,30 +0,0 @@
-module Db
-  module ProofingComponent
-    class Add
-      TOKEN_ALLOWLIST = %i[
-        address_check
-        document_check
-        document_type
-        liveness_check
-        resolution_check
-        source_check
-        verified_at
-      ].freeze
-      # address_check: lexis_nexis_phone, usps
-      # document_check: acuant
-      # document_type: state_id
-      # liveness_check: acuant
-      # resolution_check: lexis_nexis
-      # source_check: aamva
-
-      def self.call(user_id, token, value)
-        return unless user_id
-        return unless TOKEN_ALLOWLIST.index(token.to_sym)
-
-        proofing_cost = ::ProofingComponent.create_or_find_by(user_id: user_id)
```

Comment:
So arguably changing this to the `user.proofing_component || user.build_proofing_component` is more sensitive to race conditions than this existing code, I was not sure how to leverage the associations to do this same behavior

Reply:
Another option: `user.proofing_component || user.create_proofing_component` which shortens the time between queries

Reply:
I'm not entirely sure I follow what practical impact we might expect from building vs. creating, but to your alternative, maybe we don't bother `update`-ing at all if the entry doesn't exist?

```rb
if current_user.proofing_component
  current_user.proofing_component.update(component_attributes)
else
  current_user.create_proofing_component(component_attributes)
end
```

Or assign with an `upsert` ?

```rb
current.user.proofing_component = ProofingComponent.upsert(
  user: current_user,
  **component_attributes,
)
```

Reply:
The previous implementation was resilient to a race condition where two conflicting records were created at the same time, and I think would be fine in the update case since we're only typically updating a column or two at a time. The current implementation I don't think accounts for either case.

Upsert would make the most sense imo both for performance and race-condition reasons.

Reply:
I like that idea!

A few small snags: 

- from the [upsert docs](https://apidock.com/rails/v6.0.0/ActiveRecord/Persistence/ClassMethods/upsert), it skips ActiveRecord callbacks, which means no `updated_at` and no `created_at`
```ruby
[3] pry(main)> ProofingComponent.upsert(user_id: 1000)
D, [2021-12-21T09:25:36.713454 #44575] DEBUG -- :   ProofingComponent Upsert (16.5ms)  INSERT INTO "proofing_components" ("user_id") VALUES (1000) ON CONFLICT ("id") DO UPDATE SET updated_at=(CASE WHEN ("proofing_components"."user_id" IS NOT DISTINCT FROM excluded."user_id") THEN "proofing_components".updated_at ELSE CURRENT_TIMESTAMP END),"user_id"=excluded."user_id" RETURNING "id"
ActiveRecord::NotNullViolation: PG::NotNullViolation: ERROR:  null value in column "created_at" of relation "proofing_components" violates not-null constraint
DETAIL:  Failing row contains (6, 1000, null, null, null, null, null, null, null, null, null).
```
- passing `created_at` to `upsert` would override an existing created_at which we don't want...
- it doesn't return an AR model, so the `current_user.proofing_component = ` part wouldn't work but we don't really need that

Reply:
I'm going to keep noodling on this and see if I can find a way to do what we want

Reply:
updated in b238e995d0413011f37d902194e54ffa2c3acd7b

I had a few ideas:
- switch the assocation to `has_many :proofing_components` (even though there's only one) because then we can do `user.proofing_components.first_or_create`
- unfortunately that is not the same as `create_or_find_by` which still require the `user` to be passed in, so I just went with that